nrf_modem: doc: refactor RAI (#33)
lemrey/sdk-nrfxlib: nrf_modem/doc/rai.rst

.. _release_assistance_indication:

> Release Assistance Indication (RAI)

Reviewer: I don't think this page should be at this level, It should be part of `sockets.rst` page instead.
This level is for high-level library features, like interfaces, tracing etc.

Reviewer 2: I agree with Emanuele's comment. I had this thought earlier, but since the document has a lot of content, we could have a reference from line no:164 of the socket.rst doc to this page?

Author: Good point, however the `sockets.rst` file is expanding constantly and it may become increasingly difficult to maintain. What do you think @divipillai about the creation of a `sockets` subdirectory and maintaining `sockets.rst` with references to subdocuments contained in the new folder? Features that could go there could be for example `PDN` and `RAI` for now.

Reviewer 2: Agree with you. This is something followed here also: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/asset_tracker_v2/README.html. We have one more layer of subpages for Application module: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/asset_tracker_v2/doc/modules.html.

Author: @divipillai I'm opening a task for moving `RAI` and `PDN` into a `sockets` subdirectory. In the meantime, for this commit, I will merge `rai.rst` into `sockets.rst`.

###################################

.. contents::
   :local:
   :depth: 2

Introduction
************

Release Assistance Indication (RAI) is a feature introduced in 3GPP Release 13 and further enhanced in Release 14. It is designed to optimize the power consumption of IoT devices by allowing them to inform the network that they do not expect any more data to be sent or received after their current transmission. This enables the device to transition to a low-power state more quickly.
RAI is particularly useful for devices operating on LTE-M and NB-IoT networks, where power efficiency is crucial. By using RAI, devices can reduce the time they spend in the higher power-consuming RRC_CONNECTED state, thus saving battery life.

.. note::

>    It is important to note that the decision to release the connection is still at the network's discretion.

Author: @divipillai `decision` and `decided` are redundant. `Discretion` in the sentence conveys precisely the meaning while also being widely used within NCS and Zephyr's documentation. Was the goal of this change to simplify the vocabulary?

>    It is important to note that the decision to release the connection is still at the network's discretion.

Reviewer 2: Yeah, that was the intention. I haven't see "Discretion" used in the NCS documentation. We don't follow the zephyr documentation style though.

>    It is important to note that the decision to release the connection is still at the network's discretion.

Reviewer 2: Checked with one of my colleague, Pekka, and as per him : It is grammatically correct but it sort of personifies network. So he suggested the following :
_"The network has control over the connection. 
  It may or may not release the connection."_

We could use this.

>    It is important to note that the decision to release the connection is still at the network's discretion.

Author: In this context `network` is the network provider rather than network as a network of entities. I think that's it's acceptable to use `discretion` in this case.

Perhaps:

` The decision to release the connection is still at the discretion of the network provider`

>    It is important to note that the decision to release the connection is still at the network's discretion.

Reviewer 2: Agree with this since this is in context with the network provider. Saw similar example in the infocenter docs: https://infocenter.nordicsemi.com/index.jsp?topic=%2Fstruct_nrftools%2Fstruct%2Fnrftools_nrfconnect.html

There are two types of RAI:

> - **Control Plane RAI (CP-RAI)**: Used for control plane data in NB-IoT.
> - **Access Stratum RAI (AS-RAI)**: Used for both control plane and user plane data in LTE-M and NB-IoT.

Reviewer 2: ```suggestion
* Control Plane RAI (CP-RAI) - Used for control plane data in NB-IoT.
* Access Stratum RAI (AS-RAI) -  Used for both control plane and user plane data in LTE-M and NB-IoT.
```
Following the style used in NCS.

Advantages of RAI
*****************

> 
> - **Reduced Power Consumption**: By avoiding unnecessary RRC connection states, devices conserve power.
> - **Improved Battery Life**: Extends the operational lifespan of battery-powered IoT devices.
> - **Enhanced Network Efficiency**: Helps in reducing network congestion by minimizing the time devices spend connected to the network.

Reviewer 2: ```suggestion
The Following are the advantages of using RAI:

* Reduced power consumption - Devices conserve power by avoiding unnecessary RRC connection states.
* Improved battery life - Extends the operational lifespan of battery-powered IoT devices.
* Enhanced network efficiency - Helps in reducing network congestion by minimizing the time devices spend connected to the network.
```

Enabling RAI
************

> RAI can be enabled or disabled using the AT command interface:
> 
> - Enable: ``AT%RAI=1``
> - Disable: ``AT%RAI=0``

Reviewer 2: ```suggestion
RAI can be enabled or disabled using the following AT command interface:

* To enable: ``AT%RAI=1``
* To disable: ``AT%RAI=0``
```

Author: @divipillai `the following AT command interface:` would be incorrect. I can update it to: `the following AT commands:`.

Reviewer 2: Agree.

Using RAI
*********

> After enabling RAI, you control it using the ``nrf_setsockopt`` function with the ``NRF_SO_RAI`` socket option and with the following values:
> 
> - ``NRF_RAI_NO_DATA``: No more outgoing data and no incoming data expected.
> - ``NRF_RAI_LAST``: No more outgoing data after the next send operation.
> - ``NRF_RAI_ONE_RESP``: Expecting a single response after the next send operation.
> - ``NRF_RAI_ONGOING``: The socket is actively used, and the connection should be maintained.
> - ``NRF_RAI_WAIT_MORE``: The socket is actively used by a server application, and the connection should be maintained.

Reviewer 2: ```suggestion
After enabling RAI, you control it using the :c:func:`nrf_setsockopt` function with the :c:macro:`NRF_SO_RAI` socket option and with the following values:

* :c:macro:`NRF_RAI_NO_DATA` - No more outgoing data and no incoming data are expected.
* :c:macro:`NRF_RAI_LAST` - No more outgoing data after the next send operation.
* :c:macro:`NRF_RAI_ONE_RESP` - Expecting a single response after the next send operation.
* :c:macro:`NRF_RAI_ONGOING` - The socket is actively used, and the connection should be maintained.
* :c:macro:`NRF_RAI_WAIT_MORE` - The socket is actively used by a server application, and the connection should be maintained.
```

> Code Examples
> *************
> 
> The following code snippets illustrate some cases on how to apply RAI using the ``nrf_setsockopt()`` function.
> 
> Example: Sending a UDP Packet with RAI_LAST
> -------------------------------------------

Reviewer 2: ```suggestion
Code examples
=============

The following code snippets illustrate some cases on how to apply RAI using the :c:func:`nrf_setsockopt` function.

For an example for sending a UDP Packet with :c:macro:`NRF_RAI_LAST`, see the following code snippet:
```

Author: @divipillai `For an example for sending a UDP Packet with :c:macro:`NRF_RAI_LAST`, see the following code snippet:` is particularly verbose. Is there any way this could be simplified?

.. code-block:: c

   #include <nrf_socket.h>

   int send_udp_packet_with_rai_last(int fd)
   {
       int err;
       const char message[] = "Hello, World!";
       struct nrf_sockaddr_in dest_addr = {
           .sin_family = NRF_AF_INET,
           .sin_port = nrf_htons(1234),
           .sin_addr = {
               .s_addr = nrf_htonl(NRF_INADDR_ANY)
           }
       };

       /* Set the RAI_LAST option to indicate that no further data will be sent after this packet */
       int option = NRF_RAI_LAST;
       err = nrf_setsockopt(fd, NRF_SOL_SOCKET, NRF_SO_RAI, &option, sizeof(option));
       if (err) {
           printf("Failed to set NRF_SO_RAI option, error: %d\n", err);
           return err;
       }

       /* Transmit the message */
       err = nrf_sendto(fd, message, sizeof(message), 0,
                        (struct nrf_sockaddr *)&dest_addr, sizeof(dest_addr));
       if (err < 0) {
           printf("Failed to send UDP packet with RAI_LAST, error: %d\n", err);
           return err;
       }

       return 0;
   }

> Example: Receiving a UDP Packet and Setting RAI_NO_DATA
> -------------------------------------------------------

Reviewer 2: ```suggestion
For an example for receiving a UDP Packet and Setting with :c:macro:`NRF_RAI_LAST` , see the following code snippet:
```

.. code-block:: c

   #include <nrf_socket.h>

   int receive_udp_packet_and_set_rai_no_data(int fd)
   {
       int err;
       char buffer[128];
       struct nrf_sockaddr_in src_addr;
       nrf_socklen_t addrlen = sizeof(src_addr);

       /* Receive an incoming message */
       err = nrf_recvfrom(fd, buffer, sizeof(buffer), 0,
                          (struct nrf_sockaddr *)&src_addr, &addrlen);
       if (err < 0) {
           printf("Failed to receive UDP packet, error: %d\n", err);
           return err;
       }

       /* Set the RAI_NO_DATA option to signal that the device does not expect to send or receive further data */
       int option = NRF_RAI_NO_DATA;
       err = nrf_setsockopt(fd, NRF_SOL_SOCKET, NRF_SO_RAI, &option, sizeof(option));
       if (err) {
           printf("Failed to set NRF_SO_RAI option, error: %d\n", err);
           return err;
       }

       return 0;
   }

> .. note::
> 
>    The examples assume network support for both CP-RAI and AS-RAI.
> 
> .. note::
> 
>    RAI flags can be used for secure sockets similarly to other sockets.
> 
> .. note::
> 
>    Implementing RAI should be a strategic decision based on the application's knowledge of its data transmission patterns. If there is uncertainty regarding future data transfers, it is advisable to refrain from using RAI. This caution helps avoid the additional energy expenditure associated with re-establishing the radio connection.

Reviewer 2: ```suggestion
These examples assume network support for both CP-RAI and AS-RAI.
RAI flags can be used for secure sockets similarly to other sockets.

.. note::
   Implementing RAI should be a strategic decision based on the application's knowledge of its data transmission patterns. 
   If there is uncertainty regarding future data transfers, it is advisable to refrain from using RAI. 
   This helps to avoid the additional energy expenditure associated with re-establishing the radio connection.
```
The 1st two notes don't fall under the note admonitions. https://nordicsemi.atlassian.net/wiki/spaces/TECHDOC/pages/120306494/Style+and+usage+guide

Author: @divipillai I agree. According to the `style and usage guide` it should fall under `important`. The reason is that the code example will not show the expected behavior if the network doesn't support the feature.

Checking RAI Status
*******************

> To verify if the RAI information was reported to the network:
> 
> - For CP-RAI, check the ESM DATA TRANSPORT PDU sent by the modem.
> - For AS-RAI, currently, it is not visible in customer builds without specific modem traces.

Reviewer 2: ```suggestion
To verify if the RAI information was reported to the network:

* For CP-RAI, check the ESM DATA TRANSPORT PDU sent by the modem.
* For AS-RAI, currently, it is not visible in customer builds without specific modem traces.
```

RAI Best Practices
******************

> - Use RAI when no additional uplink or downlink traffic is expected in the near term.
> - Avoid using RAI if subsequent data transmission is anticipated, as it may lead to unnecessary RRC bearer reestablishments.
> - Consider each socket separately for RAI usage, and the modem will aggregate the RAI status across all sockets.

Reviewer 2: ```suggestion
Following are the best practices to be considered while using RAI:

* Use RAI when no additional uplink or downlink traffic is expected in the near term.
* Avoid using RAI if subsequent data transmission is anticipated, as it may lead to unnecessary RRC bearer reestablishments.
* Consider each socket separately for RAI usage, and the modem will aggregate the RAI status across all sockets.
```

> Relevant Links
> **************

Reviewer 2: ```suggestion
References
**********
```

> - `3GPP TS 24.301 <https://www.3gpp.org/DynaReport/24301.htm>`_

Reviewer 2: The link needs to be added to link.txt.

Conclusion
**********

> RAI is a useful feature for optimizing network resource usage and device power consumption. By indicating the end of data transmission, devices can potentially reduce the time they spend connected to the network, saving battery life and freeing up network resources.
> Remember to test the RAI functionality with your network provider to ensure compatibility and to understand the network's behavior in response to RAI signals.

Reviewer 2: ```suggestion
RAI is a useful feature for optimizing network resource usage and device power consumption. 
By indicating the end of data transmission, devices can potentially reduce the time they spend connected to the network, saving battery life and freeing up network resources.
The RAI functionality must be tested with your network provider to ensure compatibility and to understand the network's behavior in response to RAI signals.
```
